lib/test/Test/Pos/Ssc/Toss/BaseSpec.hs

```diff
@@ -594,28 +601,53 @@ checksBadCertsPayload (GoodPayload epoch sgs certsMap mrs) pk cert =
-            c2id <- head (HM.keys (getVssCertificatesMap certsMap))
-            let certsMap4 = certsMap
-                    & _Wrapped . ix c2id . _vcVssKey .~ vcVssKey c1
-                certDuplicateVss =
-                    tossRunner mrs sgs $
-                    checkCertificatesPayload epoch certsMap4
-            pure $ c1id /= c2id ==> case certDuplicateVss of
-                Left (CertificateDuplicateVssKey nes) -> c2id `qcElem` nes
-                _ -> qcFail $ "expected " <> show certDuplicateVss <>
-                              " to be a Left (CertificateDuplicateVssKey ...)"
-
-        allVssKeys = map vcVssKey (toList (getVssCertificatesMap certsMap))
+        getMapVssKeys :: VssCertificatesMap -> [AsBinary VssPublicKey]
+        getMapVssKeys = map vcVssKey . HM.elems . getVssCertificatesMap
+
+        certsMapVssKeys = getMapVssKeys certsMap
+
+        -- [CT-99]: VssCertificatesMap should not be allowed to contain a
+        -- certificate with a VSS key that already exists in VssCertData.
+        res4 = do let vssDataCertsMap = certs $ sgs ^. sgsVssCertificates
+                  let dataKeys  = getMapVssKeys vssDataCertsMap
+                  let anyInData = any (`elem` dataKeys) certsMapVssKeys -- is there key from map in VssData?
+                  not anyInData
 
     in (not (HM.member certSid $ mrs HM.! epoch) &&
-        not (vcVssKey cert `elem` allVssKeys))
+        not (vcVssKey cert `elem` certsMapVssKeys))
        ==> res1 .&&. res2 .&&. res3 .&&. res4
 
+-- [CT-99]: Test that no two stakeholders can have same active VSS keys. Specifically:
+--
+-- A VSS payload with two certs that have equal VSS keys should not be
+-- deserializable (with decode).
+checksBijectiveCertsPayload :: HasConfiguration
+                            => VssCertificatesMap
+                            -> SecretKey
+                            -> SecretKey
+                            -> VssPublicKey
+                            -> Property
+checksBijectiveCertsPayload (UnsafeVssCertificatesMap vssCertsMap) sk1 sk2 vssPk = do
+    -- generate two certificates with different SecretKeys but same VssKey
+    let binaryVssPk = asBinary vssPk
+    let cert1 = mkVssCertificate sk1 binaryVssPk 0
+    let cert2 = mkVssCertificate sk2 binaryVssPk 0
+    let key1  = addressHash (vcSigningKey cert1)
+    let key2  = addressHash (vcSigningKey cert2)
+    let newCerts = UnsafeVssCertificatesMap
+                 $ HM.insert key1 cert1
+                 $ HM.insert key2 cert2 vssCertsMap
+
+    -- check that certificate duplication throws error on decoding stage
+    let encodedCerts = encode newCerts
+    let decodedCerts = deserialiseFromBytes (decode @VssCertificatesMap)
+                                            (toLazyByteString encodedCerts)
+    let check = case decodedCerts of
+            Left _  -> property True
+            Right _ -> qcFail "Decoding of VssCertMap with duplicate certs should fail"
+
+    sk1 /= sk2 ==> check
```

Comment:
This one looks good.

Reply:
Looks fine from my side

Reply:
Note that this is no longer true in the develop branch I think. These kinds of validation tests are being moved out of deserialisation and into more appropriate places. Worth checking if this is the case for this validation check.